crates/block-producer/src/block_builder/prover/block_witness.rs

```diff
@@ -37,37 +37,62 @@ impl BlockWitness {
         Self::validate_inputs(&block_inputs, batches)?;
 
         let updated_accounts = {
-            let mut account_initial_states: BTreeMap<AccountId, Digest> =
-                batches.iter().flat_map(TransactionBatch::account_initial_states).collect();
+            // Each accounts initial state in the context of the block i.e. only the first
+            // transaction affecting an account matters.
+            let mut account_initial_states = BTreeMap::<AccountId, Digest>::new();
+            for (account_id, initial_state) in
+                batches.iter().flat_map(TransactionBatch::account_initial_states)
+            {
+                account_initial_states.entry(account_id).or_insert(initial_state);
+            }
```

Comment:
I think the assumption we make here is that account updates across batches are ordered chronologically. That is, if the same account is updated from `a -> b` in one batch and from `b -> c` in another. The batch with `a -> b` update comes first - right?

If so, I think we should state this assumption explicitly, and maybe in the future we can relax this condition somehow because batches are not necessarily meant to be sequential.

Reply:
Ah. I'll try fix that and otherwise just add a comment.

Reply:
I think it is important to note down somewhere that:

- We currently expect transactions in a batch to be ordered chronologically.
- We do not expect the batches in the block to be ordered chronologically.

Though I think a good question would be why the difference? I'm not sure I have a good answer to this yet :)

Reply:
I guess maybe one justification:

It should be pretty easy to order transactions chronologically for a batch. It may not be so easy (or may be even impossible) to order batches in the same way in the block.

For example, let's say we have accounts `a` and `b`. Batch 1 updates account `a` from `a0 -> a1` and account `b` from `b1 -> b2`. Batch 2 updates account `a` from `a1 -> a2` and account `b` from `b0 -> b1`. Not sure if such situation can arise in practice, but theoretically, it should be possible. And in this case, two batches cannot be ordered chronologically with respect to each other.

Reply:
My (attempt at a) justification:

A transaction is rejected if it is not chronologically ordered i.e. it is rejected if the nonce and state transitions are not correct with the latest state on the node. Transactions are therefore placed into batches in chronological order.

There are many possible strategies for placing transactions into batches e.g.
- merge as many transactions together (compression)
- delay some transactions for MEV purposes
- minimize coupling between batches

A very simple batch strategy might just be round-robin style. This can result in a cyclic coupling between batches which means they cannot be ordered. Example: two batches `b`, two accounts `x`,`y`, with two transactions each `x0`, `x1`, `y0`, `y1`.

If the transactions arrive as `x0`,`y0`,`y1`,`x1` then we get batches:
```
b0 = [x0,y1]
b1 = [y0,x1]
```
which means there is no way to chronologically order the batches. This could be disallowed by tracking a batch dependency graph -- and there might be an argument that we need this anyway so that if a batch fails, we know which other batches must wait for it to be rebuilt.

Reply:
Put differently: it is already illegal for a batch to contain non-sequential transactions for the same account. It might be simplest to also disallow inter-account cycles.

Reply:
Yes - agreed. In terms of how to implement this, one option is to do the following:

1. For each account, get the initial state from `BlockInputs`. Set the latest known state of this account to this value.
2. Iterate over all batches and find a batch where the account's initial state and the latest known sate match. Then set the latest known state to the final account state from this batch.
3. Repeat the above process until no batches have the initial state of the account match the current latest state.

This is probably not super efficient (and there could be ways to improve on this) - but I think it should work.

Reply:
I was thinking of that; but it would involve some emptiness checks and care to avoid leaving something behind.

What I'm going to try is to first collect all batches into a nested map: `Map<AccountId, Map<InitState, Update>>`. I can then start at the accounts input state and iterate through the inner update map until done.

I had started doing this; but then realised I should be doing this merging in the block updater instead - since it has similar issues of assuming batches are independent (and also invokes the witness builder).